Contrib/FreeWilson/freewilson.py

```diff
@@ -0,0 +1,486 @@
+        self.rgroup_to_descriptor_idx = rgroup_to_descriptor_idx # dictionary {smi:descriptor_idx}
+        self.fitter = fitter # fitter rgroup indices -> prediction
+        self.N = len(rgroup_to_descriptor_idx)
+        self.r2 = r2
+        self.descriptors = set([tuple(d) for d in descriptors])
+
+
+default_decomp_params = rgd.RGroupDecompositionParameters()
+
+# The default decomposition uses the GeneticAlgorirthm
+#   fingerprint analysis to break symmetry and make
+#   more consistent rgroup decompositions
+default_decomp_params.matchingStrategy = rgd.GA
+
+# Also the decomposition is allowed to add new rgroups
+default_decomp_params.onlyMatchAtRGroups = False
+
+def FWDecompose(scaffolds, mols, scores, decomp_params=default_decomp_params) -> FreeWilsonDecomposition:
+    """
+    Perform a free wilson analysis
+        : param scaffolds : scaffold or list of scaffolds to use for the rgroup decomposition
+        : param mols : molecules to decompose
+        : param scores : list of floating point numbers for the regression (
+                             you may need convert these to their logs in some cases)
+        : param decomp_params : RgroupDecompositionParams default [
+                                    default_decomp_params = rdkit.Chem.rdRGroupDecomposition.RGroupDecompositionParameters()
+                                    default_decomp_params.matchingStrategy = rgd.GA
+                                    default_decomp_params.onlyMatchAtRGroups = False
+                                   ]
+                                If you only want to decompose on specific group locations
+                                set onlyMatchAtRGroups to True
+
+
+        >>> from rdkit import Chem
+        >>> from rdkit.Chem import Descriptors, FreeWilson
+        >>> scaffold = Chem.MolFromSmiles("c1cccnc1")
+        >>> mols = [Chem.MolFromSmiles("c1cccnc1"+"C"*(i+1)) for i in range(100)]
+        >>> scores = [Descriptors.MolLogP(m) for m in mols]
+        >>> fw = FWDecompose(scaffold, mols, scores)
+        >>> for pred in FWBuild(fw):
+        >>>  ...
+
+    To filter out molecules in the training set, send the freewilson decomoposition into
+    the builer.
+
+       >>> for pred in FWBuild(fw, decomposition=fw):
+       >>>    ...
+
+    For an easy way to report predictions see 
+
+       >>> import sys
+       >>> predictions_to_csv(sys.stdout, fw, FWBuild(fw))
+
+   
+    See FWBuild docs to see how to filter predictions, molecular weight or molecular properties.
+    """
+    descriptors = descriptors = [] # list of descriptors, one per matched molecules
+                                   #    descriptors are 1/0 if a sidechain is present
+    matched_scores = []            # scores from the matching molecules
+    rgroup_idx = rgroup_idx = {}   # rgroup index into descriptor { smiles: idx }
+    rgroups = rgroups = defaultdict(list) # final list of rgrups/sidechains
```

Comment:
```suggestion
    descriptors = [] # list of descriptors, one per matched molecules
                                   #    descriptors are 1/0 if a sidechain is present
    matched_scores = []            # scores from the matching molecules
    rgroup_idx = {}   # rgroup index into descriptor { smiles: idx }
    rgroups = defaultdict(list) # final list of rgrups/sidechains
```

Reply:
Good eye, these were carried over from moving this into a class.